applications/nrf_desktop/src/modules/hid_forward.c

```diff
@@ -716,29 +734,25 @@ static void disconnect_peripheral(struct hids_peripheral *per)
 	}
 }
 
-static void clear_state(void)
+static void enable_notifications(struct subscriber *sub, uint8_t report_id)
 {
-	for (size_t per_id = 0; per_id < ARRAY_SIZE(peripherals); per_id++) {
-		struct hids_peripheral *per = &peripherals[per_id];
-
-		for (size_t report_id = 0; report_id < ARRAY_SIZE(per->enqueued_report); report_id++) {
-			struct enqueued_reports *enqueued_report = &per->enqueued_report[report_id];
-
-			/* Clear all the reports. */
-			while (!sys_slist_is_empty(&enqueued_report->list)) {
-				struct enqueued_report *item;;
+	sub->enabled_reports_bm |= BIT(report_id);
+}
 
-				item = CONTAINER_OF(sys_slist_get(&enqueued_report->list),
-						     __typeof__(*item),
-						    node);
-				enqueued_report->count--;
+static void disable_notifications(struct subscriber *sub, uint8_t report_id)
+{
+	sub->enabled_reports_bm &= ~BIT(report_id);
 
-				k_free(item->report);
-				k_free(item);
-			}
+	/* For all peripherals connected to this subscriber. */
+	for (size_t per_id = 0; per_id < ARRAY_SIZE(peripherals); per_id++) {
+		struct hids_peripheral *per = &peripherals[per_id];
 
-			__ASSERT_NO_MSG(enqueued_report->count == 0);
+		if ((sub != get_subscriber(per)) ||
+		    (!is_peripheral_connected(per))) {
```

Comment:
I think we should drop the enqueued reports also in case the peripheral is not connected.

Reply:
Then we may generate a report with all keys released when subscriptions are enabled to properly clear pressed keys.

Reply:
First check allows reports to be temporarily enqueued when USB transition state. I would not be against removing it but that would have to go along with removing enqueueing while sub is not connected (when report is forwarded). This part was added because of USB suspend. I am not sure it has much value as USB goes through the disabled state always so we could assume host does not care to have continuous reception of reports here. I will check if this can be added to the PR.

Reply:
Regarding the second part, the idea is good but cannot be implemented with current report enqueueing.
To enqueue the report we need to have a peripheral connected to sub. We could zero out the reports when peripheral connect but I seen no point in it as each connected peripheral does it on its own.

Reply:
I have reworked the enqueueing so we are now able to drain the reports from sub instead of per level.